File: gradle-package/src/main/java/io/github/tejasmk2/gradle/calculator/Calculator.java
package io.github.tejasmk2.gradle.calculator;

import java.math.BigDecimal;
import java.math.MathContext;
import java.util.Objects;

/** Robust decimal arithmetic engine for Gradle applications. */
public final class Calculator {
    private static final MathContext MC = MathContext.DECIMAL128;

    private Calculator() {
    }

    public static BigDecimal add(BigDecimal a, BigDecimal b) { return require(a).add(require(b), MC); }
    public static BigDecimal subtract(BigDecimal a, BigDecimal b) { return require(a).subtract(require(b), MC); }
    public static BigDecimal multiply(BigDecimal a, BigDecimal b) { return require(a).multiply(require(b), MC); }

    public static BigDecimal divide(BigDecimal a, BigDecimal b) {
        require(a); require(b);
        if (b.signum() == 0) throw new ArithmeticException("cannot divide by zero");
        return a.divide(b, MC);
    }

    public static BigDecimal modulo(BigDecimal a, BigDecimal b) {
        require(a); require(b);
        if (b.signum() == 0) throw new ArithmeticException("cannot modulo by zero");
        return a.remainder(b, MC);
    }

    public static BigDecimal power(BigDecimal a, int exponent) { return require(a).pow(exponent, MC); }

    public static BigDecimal percentage(BigDecimal value, BigDecimal percent) {
        return require(value).multiply(require(percent), MC).divide(BigDecimal.valueOf(100), MC);
    }

    private static BigDecimal require(BigDecimal value) {
        return Objects.requireNonNull(value, "value must not be null");
    }
}

File: nuget-package/Calculator.cs
namespace Pijush.Calculator;

/// <summary>Robust decimal arithmetic engine for .NET applications.</summary>
public static class Calculator
{
    public static decimal Add(decimal a, decimal b) => a + b;
    public static decimal Subtract(decimal a, decimal b) => a - b;
    public static decimal Multiply(decimal a, decimal b) => a * b;

    public static decimal Divide(decimal a, decimal b)
    {
        if (b == 0m) throw new DivideByZeroException("Cannot divide by zero.");
        return a / b;
    }

    public static decimal Modulo(decimal a, decimal b)
    {
        if (b == 0m) throw new DivideByZeroException("Cannot modulo by zero.");
        return a % b;
    }

    public static decimal Power(decimal value, int exponent)
    {
        if (exponent == 0) return 1m;
        var negative = exponent < 0;
        var power = Math.Abs((long)exponent);
        var result = 1m;
        var baseValue = value;
        while (power > 0)
        {
            if ((power & 1) == 1) result *= baseValue;
            power >>= 1;
            if (power > 0) baseValue *= baseValue;
        }
        return negative ? 1m / result : result;
    }

    public static decimal Percentage(decimal value, decimal percent) => value * percent / 100m;
}
